here-naksha-lib-model/src/jvmTest/kotlin/naksha/model/PropertyFilterTest.kt

```diff
@@ -0,0 +1,161 @@
+package naksha.model
+
+import naksha.base.AnyObject
+import naksha.base.Int64
+import naksha.jbon.JbEncoder
+import naksha.model.objects.NakshaFeature
+import naksha.model.request.ExecutedOp
+import naksha.model.request.PropertyFilter
+import naksha.model.request.ReadFeatures
+import naksha.model.request.ResultTuple
+import naksha.model.request.query.*
+import org.junit.jupiter.api.BeforeAll
+import org.junit.jupiter.api.Test
+import org.mockito.kotlin.mock
+import kotlin.test.assertEquals
+
+class PropertyFilterTest {
+
+    companion object {
+        lateinit var resultTuple : ResultTuple
+        val nestedJson = AnyObject()
+
+        @JvmStatic
+        @BeforeAll
+        fun setupTuple(): Unit {
+            // create the feature
+            val feature = NakshaFeature()
+            feature.properties["foo"] = "bar"
+            feature.properties["number"] = 1.1
+            nestedJson["bool"] = true
+            nestedJson["nullProps"] = null
+            val innerJson = AnyObject()
+            innerJson["a"] = 1
+            nestedJson["array"] = arrayOf("one", "two", "three", innerJson)
+            feature.properties["json"] = nestedJson
+            // build tuple containing the feature
+            val encoder = JbEncoder()
+            val byteArray = encoder.buildFeatureFromMap(feature)
+            val storeNumber = StoreNumber(0, Int64(0))
+            val version = Version(0)
+            val flag = Flags(0)
+            val tupleNumber = TupleNumber(
+                storeNumber = storeNumber,
+                uid = 0,
+                version = version,
+            )
+            val mockStorage = mock<IStorage>()
+            val tuple = Tuple(
+                storage = mockStorage,
+                tupleNumber = tupleNumber,
+                Metadata(
+                    storeNumber = storeNumber,
+                    updatedAt = Int64(0),
+                    uid = 0,
+                    id = "",
+                    appId = "",
+                    author = null,
+                    version = version,
+                    type = null,
+                    flags = flag,
+                ),
+                byteArray
+            )
+            resultTuple = ResultTuple(
+                storage = mockStorage,
+                tupleNumber = tupleNumber,
+                op = ExecutedOp.READ,
+                featureId = null,
+                tuple = tuple
+            )
+        }
+    }
```

Comment:
NTH:
Consider moving creating of this `ResultTuple` out of this test . so that the test class contains mainly the test and then the setup is minimal ;) It can be a bit easier to follow up with this.
I would probably delegate this job to some util object in different file, as we did with `ProxyFeatureGenerator`.

Then, if you had such function, you can create new `ResulttTuple` for each test and supply only the parts that interest you (for example: why would you need to specify stuff like `flags` and `storage` if only that we care about in this test concerns the properties, let us only play with properties then - the rest of needed stuff can be internal logic of this generator object.

Another good thing that would come out if this is that **we use different object instance** for each test - hence, we eliminate possibility of caching/mutability/whatever breaking our tests "by accident" ;)
The costs are none, this is test environment, and creating dozens of instances is not an ussue.

Reply:
Thanks, good idea about minimizing smaller args like `flags`. It was from the time I tried mockative, which requires opening up all classes upfront, so mocking then was not a good option. But it's ok now since using mockito only for JVM.

There is one little cost when using multiple feature instances for testing though (especially when the tests are split into smaller cases like you suggested above) is that the CICD pipeline will take longer to run (all the more so if this feature generator will also be used for many other tests). But yes, the advantage would be a convenient feature generator possibly for every hub modules.

What do you think? I want to have a separated module just for testing utils, where we put this feature generator into it. The feature internal is mocked by default, unless testers supply definite values for things they want to test. Then all modules can depend on this module for convenient testing.

Reply:
First, the performance impact should be close to nothing. Creating a new object is a matter of micro or nanoseconds (you can optimize the creation later, the largest cost normally are when generating random numbers).

About an own module, I would oppose it. What may be helpful, is a module that generates test data, but not in a specific format, rather with helper methods like `randomName()`, `randomFirstName()`, `randomLatitude()`, `randomLongitude()` and so on, but always only creating abstract data objects using our `lib-base`, so we can use it really everywhere. Especially, when we later want to start testing unicode, like Chinese or other special languages. I just would say, we should make an own task out of this (so add some JIRA ticket to backlog), it may be too much right now.

Reply:
Create a task in backlog for this. For this specific test suit, I would leave it as is, and switch to using this feature generator later when we have it.